src/cohort/cohort_server.cc

```diff
@@ -1,56 +1,265 @@
+  // occur if we wait for both the read and write locks.
+  absl::flat_hash_set<std::string> readonly_keys;
+  for (const common::Operation& op : transaction.ops()) {
+    if (op.has_put()) {
+      readonly_keys.erase(op.put().key());
+      write_and_readwrite_keys.emplace(op.put().key());
+    }
+    if (op.has_get() && !write_and_readwrite_keys.contains(op.get().key())) {
+      readonly_keys.emplace(op.get().key());
+    }
+  }
+  for (const auto& read_key : readonly_keys) {
+    if (!GetLock(read_key).ReaderLockWhenWithDeadline(absl::Condition::kTrue,
+                                                      presumed_abort_time)) {
+      return absl::DeadlineExceededError(
+          absl::StrCat("Could not acquire read lock for ", read_key,
+                       " before the abort deadline"));
+    }
+    txn_metadata.read_lock_keys.push_back(read_key);
+  }
+  for (const auto& write_key : write_and_readwrite_keys) {
+    if (!GetLock(write_key).WriterLockWhenWithDeadline(absl::Condition::kTrue,
+                                                       presumed_abort_time)) {
+      return absl::DeadlineExceededError(
+          absl::StrCat("Could not acquire write lock for ", write_key,
+                       " before the abort deadline"));
+    }
+    txn_metadata.write_lock_keys.push_back(write_key);
+  }
+  return absl::OkStatus();
+}
+
+absl::Status CohortServer::ProcessOperationInDb(
+    const common::Operation& op, internal::TransactionMetadata& txn_metadata) {
+  if (op.has_get()) {
+    int64_t value;
+    RETURN_IF_ERROR(txn_metadata.db->Get(op.get().key(), value));
+    common::GetResponse* get_response =
+        txn_metadata.response.mutable_committed_response()->add_get_responses();
+    *get_response->mutable_get() = op.get();
+    *get_response->mutable_namespace_() = op.namespace_();
+    get_response->mutable_value()->set_int64_value(value);
+    return absl::OkStatus();
+  }
+  if (op.has_put()) {
+    if (op.put().value().has_constant_value()) {
+      return txn_metadata.db->Put(
+          op.put().key(), op.put().value().constant_value().int64_value());
+    }
+    int64_t value;
+    absl::Status get_status = txn_metadata.db->Get(op.put().key(), value);
+    if (get_status.code() == absl::NotFoundError("").code() &&
+        op.put().value().relative_value().has_default_value()) {
+      value = op.put().value().relative_value().default_value().int64_value();
+    } else if (!get_status.ok()) {
+      return get_status;
+    }
+    return txn_metadata.db->Put(
+        op.put().key(),
+        value +
+            op.put().value().relative_value().relative_value().int64_value());
+  }
+  // No-op.
+  return absl::OkStatus();
+}
+
+absl::Status CohortServer::ProcessTransactionInDb(
+    const common::Transaction& transaction, absl::Time presumed_abort_time,
+    internal::TransactionMetadata& txn_metadata) {
+  RETURN_IF_ERROR(
+      AcquireDbLocks(transaction, presumed_abort_time, txn_metadata));
```

Comment:
I guess the current AcquireDbLocks may not be able to work with LMDB. In practice, it will simply block at Begin if it gets called two times in a single process (even from different threads). IIUC, if two concurrent txns try to Put different keys 'A' and 'B' , the `AcquireDbLocks` will return OK and continue to start Begin for each txn, where LMDB may get the 2nd txn completely blocked. 

With this limitation from LMDB, there should simply be a single db level lock. If there is open readwrite txn (starts with Begin), db locked and blocks future readwrite txn. If the current open txns are readonly, they don't block on anything

Reply:
I'll do that in the next PR.